scripts/create-darwin-volume.sh

```diff
@@ -1,33 +1,210 @@
 }
 
-# i.e., "disk1"
+# usually "disk1"
 root_disk_identifier() {
-    diskutil info -plist / | xmllint --xpath "/plist/dict/key[text()='ParentWholeDisk']/following-sibling::string[1]/text()" -
+    # For performance (~10ms vs 280ms) I'm parsing 'diskX' from stat output
+    # (~diskXsY)--but I'm retaining the more-semantic approach since
+    # it documents intent better.
+    # /usr/sbin/diskutil info -plist / | xmllint --xpath "/plist/dict/key[text()='ParentWholeDisk']/following-sibling::string[1]/text()" -
+    #
+    local special_device
+    special_device="$(/usr/bin/stat -f "%Sd" /)"
+    echo "${special_device/s[0-9]*/}"
 }
 
-find_nix_volume() {
-    diskutil apfs list -plist "$1" | xmllint --xpath "(/plist/dict/array/dict/key[text()='Volumes']/following-sibling::array/dict/key[text()='Name']/following-sibling::string[starts-with(translate(text(),'N','n'),'nix')]/text())[1]" - 2>/dev/null || true
+# Strongly assuming we'll make a volume on the device root is on
+# But you can override NIX_VOLUME_USE_DISK to create it on some other device
+readonly NIX_VOLUME_USE_DISK="${NIX_VOLUME_USE_DISK:-$(root_disk_identifier)}"
+
+substep() {
+    printf "   %s\n" "" "- $1" "" "${@:2}"
+}
+
+# $1 = label
+volumes_labeled() {
+    xsltproc --novalid --stringparam label "$1" - <(ioreg -ra -c "AppleAPFSVolume") <<'EOF'
+<xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform" version="1.0">
+  <xsl:output method="text"/>
+  <xsl:template match="/">
+    <xsl:apply-templates select="/plist/array/dict/key[text()='IORegistryEntryName']/following-sibling::*[1][text()=$label]/.."/>
+  </xsl:template>
+  <xsl:template match="dict">
+    <xsl:apply-templates match="string" select="key[text()='BSD Name']/following-sibling::*[1]"/>
+    <xsl:text>◊</xsl:text>
+    <xsl:apply-templates match="string" select="key[text()='UUID']/following-sibling::*[1]"/>
+    <xsl:text>&#xA;</xsl:text>
+  </xsl:template>
+</xsl:stylesheet>
+EOF
+    # I cut label out of the extracted values, but here it is for reference:
+    # <xsl:apply-templates match="string" select="key[text()='IORegistryEntryName']/following-sibling::*[1]"/>
+    # <xsl:text>◊</xsl:text>
+}
+
+# $1 = volume special (i.e., disk1s7)
+right_disk() {
+    [ "${1::${#NIX_VOLUME_USE_DISK}}" = "$NIX_VOLUME_USE_DISK" ]
+}
+# $1 = volume special (i.e., disk1s7)
+right_volume() {
+    # if set, it must match; otherwise ensure it's on the right disk
+    if [ -z "$NIX_VOLUME_USE_SPECIAL" ]; then
+        if right_disk "$1"; then
+            NIX_VOLUME_USE_SPECIAL="$1" # latch on
+            return 0
+        else
+            return 1
+        fi
+    else
+        [ "$1" = "$NIX_VOLUME_USE_SPECIAL" ]
+    fi
+}
+# $1 = volume uuid
+right_uuid() {
+    # if set, it must match; otherwise allow
+    if [ -z "$NIX_VOLUME_USE_UUID" ]; then
+        NIX_VOLUME_USE_UUID="$1" # latch on
+        return 0
+    else
+        [ "$1" = "$NIX_VOLUME_USE_UUID" ]
+    fi
```

Comment:
This isn’t necessarily a clear win, but it could be simplified to

```suggestion
    [ "$1" = "${NIX_VOLUME_USE_UUID:=$1}" ]
```

Reply:
Keeping the more explicit version of this.